### PR TITLE
Import relevant changes from upstream and Mina (MSM, ASM, Field refactoring, Gadgets refactoring, BN382 curves, SNARKS opt)

### LINKED ISSUE #37: Restructuring of towered extension fields Fp4 and Fp6
As Coda we take the embedding field of an MTN4 curve as the quadratic extension 

   Fp4 = Fp2[Y] / (Y^2 - X) 

of Fp2 = Fp[X]/(X^2 - alpha), where alpha is the ``const Fp2Parameters::NONRESIDUE``, and similarly the embedding field of an MNT6 curve as 

   Fp6 = Fp2[Y]/(Y^2-X)

of Fp3=Fp[X]/(X^3 - alpha), with \alpha as ``const Fp3Parameters::NONRESIDUE``.

However, the implementations of the two towered extensions differs in style, and both 'open up' the arithmetics of their subfields Fp2, Fp3, respectively, to define multiplications by $X$ therein.
To illustrate the different styles of implementation:
- for Fp4, multiplication by Y^2 = X is done as 
```
    pub fn mul_by_nonresidue(value: &Fp2<P::Fp2Params>) -> Fp2<P::Fp2Params> {
        let new_c0 = P::Fp2Params::mul_fp_by_nonresidue(&value.c1);
        let new_c1 = value.c0;
        Fp2::new(new_c0, new_c1)
    }
```
not even referring to ``Fp2Parameters::NONRESIDUE``, whereas
- for Fp6, multiplication by $X$ is implemented as
```
    pub fn mul_by_nonresidue(value: &Fp3<P::Fp3Params>) -> Fp3<P::Fp3Params> {
        let mut res = *value;
        res.c0 = value.c2;
        res.c1 = value.c0;
        res.c2 = value.c1;
        res.c0
            .mul_assign(&<P::Fp3Params as Fp3Parameters>::NONRESIDUE);
        res
    }
```
which use ``Fp3Parameters::NONRESIDUE``, and the generic ``trait Fp6Parameters`` function 
```
   fn mul_fp3_by_nonresidue(fe: &Fp3<Self::Fp3Params>) -> Fp3<Self::Fp3Params> {
        Self::NONRESIDUE * fe
    }
``` 
is not used at all.

I propose a restructuring of the towered extension fields, choosing a generic approach which defines quadratic/cubic extensions F2/F3 over any field F (not necessarily prime), and then define F6 with "sextic" non-residue ``Fp6Params::NONRESIDUE`` as quadratic extension F2(F3), using the (quadratic) non-residue  

   (c_0,c_1)=(0,1) from F3^2
 
(which corresponds to X), where F3 is the cubic extension of Fp using the sextic non-residue as (cubic) non-residue. Likewise, one defines F4 with "quartic" non-residue ``Fp4Params::NONRESIDUE`` as quadratic extension F2(F2), using the quadratic nonresidue 

   (0,1) in F2^2,

where F2 is the quadratic extension of Fp using the quartic non-residue as quadratic one.

Such generic implementation is not just for readability - it also makes further towering (if ever needed lateron) more transparent and easy.

### LINKED ISSUE #50: Optimizing deg-3 extension field gadgets
Some code of the cubic extension field gadgets `Fp3Gadget` and `Fp6Gadget` as '3over2' can be optimized:
- the `inverse` gadget uses Karatsuba multiplication using 6 base field constraints instead of Toom-Cook's 5 constraints
- the `square` gadget uses Toom-Cook multiplication, but according to [Devigili, et al.,](https://www.iacr.org/cryptodb/data/paper.php?pubkey=21962) the Chung-Hasan asymmetric squaring formula (CH-SQR2) replaces 3 multiplications by squares ([Zexe replaced this piece of code today](https://github.com/scipr-lab/zexe/pull/174/files)).
- `mul_by_constant` uses for no reason Toom-Cook with more than 100 lines, although the naive multiplication yields the same result.

Some of these issues are described in more detail in the docu branch of the field gadgets, [PR of docu field gadgets](https://github.com/ZencashOfficial/ginger-lib/pull/49).

### LINKED ISSUE #51: NEqGadget implementation of extension field gadgets
In all our extension fields gadgets, the 'not-equal' gadget `NEqGadget` as implemented enforces inequality of *all* components (=vector coordinates), and not of at least one.

Besides that this does not describe the 'canoncial' behaviour one would expect, it is necessary to evaluate the impact of using this stricter notion of inequality in other gadgets (e.g., for affine EC arithmetics, which need to compare against the 0 element, etc.).

**Update** 
The impact is surprisingly harmless. The AffineGadgets which use `enforce_not_equal` of field elements to implement the inequality gadget of points is only used at a single place of code by now (in the G2PreparedGadget of BLS12 pairings), which has no security impact.

*Still to prevent malicious impact I recommend to correct the extension field NEqGadgets to reflect the expected behaviour.*

### PR DESCRIPTION
**Pull request by 95DDB**
This PR import some relevant features and refactoring Zexe/Mina did, mainly regarding MSM, ASM and Field arithmetic. 
Also closes #37 , #50, #51

### CODE REVIEW

File: algebra/src/curves/bn_382/g.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
+use crate::{
```
Comment (UlrichHaboeck75) [inline algebra/src/curves/bn_382/g.rs:1]: I'd opt to put the dual BN-382 in a separate folder. It is a curve on its own, and the fact that it forms a cycle with the BN does not justify the same folder (we don't do that with the MNT either).

File: algebra/src/fields/models/cubic_extension.rs [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+
+use crate::{
+    bytes::{FromBytes, ToBytes},
+    bits::{FromBits, ToBits},
+    fields::{Field, FpParameters, PrimeField},
+    UniformRand, Error,
+};
+
+
+/// Model for cubic extension field of a prime field F=BasePrimeField
```
Comment (UlrichHaboeck75) [inline algebra/src/fields/models/cubic_extension.rs:16]: wrong comment here, should be cubic extension of F=BaseField (which again is an extension of BasePrimeField).

File: algebra/src/fields/models/quadratic_extension.rs [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+
+use crate::{
+    bytes::{FromBytes, ToBytes},
+    bits::{FromBits, ToBits},
+    fields::{Field, FpParameters, LegendreSymbol, PrimeField, SquareRootField},
+    UniformRand, Error,
+};
+use crate::biginteger::arithmetic::find_wnaf;
+
+/// Model for quadratic extension field of prime field F=Fp
```
Comment (UlrichHaboeck75) [inline algebra/src/fields/models/quadratic_extension.rs:16]: Also here, this is the model for quadratic extensions of a BaseField, which is again an extension of BasePrimeField.

File: algebra/src/msm/variable_base.rs [hunk 4]
```diff
@@ -1,12 +1,90 @@
...
-    AffineCurve, BigInteger, Field, FpParameters, PrimeField,
-    ProjectiveCurve,
-};
+use crate::{AffineCurve, BigInteger, Field, FpParameters, PrimeField, ProjectiveCurve};
 use rayon::prelude::*;
 
 pub struct VariableBaseMSM;
 
 impl VariableBaseMSM {
+    pub fn multi_scalar_mul_affine<G: AffineCurve>(
```
Comment (UlrichHaboeck75) [inline algebra/src/msm/variable_base.rs:7]: This function seems to be very similar to msm_affine. Let's bench one versus the other.

**Review by UlrichHaboeck75 — APPROVED** [review]
At some point in time we should carry over the extensive documentation in #49 , #62, #63, #65  to the new structure.

**Review by albertog78 — APPROVED** [review]

**Pull request merged**